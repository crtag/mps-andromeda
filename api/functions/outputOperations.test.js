// molecule-parser.test.js
const fs = require("fs");
const path = require("path");
const {extractMoleculeInput, extractSimulationResults} = require("./outputOperations");
const {describe, test, expect, beforeAll} = require("@jest/globals");

// Helper function to load test files
function loadTestFile(filename) {
    return fs.readFileSync(path.join(__dirname, "test-files", filename), "utf8");
}

describe("Molecule Parser Tests", () => {
    let successTestFileContent;

    beforeAll(() => {
        // Load your test file
        successTestFileContent = loadTestFile("quick-success.out");
    });

    describe("extractMoleculeInput", () => {
        test("extracts correct atom number and electrons from real test file", () => {
            const result = extractMoleculeInput(successTestFileContent);
            expect(result).toMatchSnapshot();
        });

        test("returns null values for invalid input", () => {
            const result = extractMoleculeInput("invalid content");
            expect(result).toEqual({
                totalAtomNumber: null,
                totalElectrons: null,
            });
        });

        test("handles empty input", () => {
            const result = extractMoleculeInput("");
            expect(result).toEqual({
                totalAtomNumber: null,
                totalElectrons: null,
            });
        });
    });
});

describe("extractSimulationResults", () => {
    let successTestFileContent;

    beforeAll(() => {
        // Load your test file
        successTestFileContent = loadTestFile("quick-success.out");
    });

    test("extracts simulation results from real test file", () => {
        const result = extractSimulationResults(successTestFileContent);
        expect(result).toMatchSnapshot();
    });

    test("returns null values for invalid input", () => {
        const result = extractSimulationResults("invalid content");
        expect(result).toEqual({
            optimizedGeometry: null,
            minimizedEnergy: null,
            totalTime: null,
        });
    });

    test("handles empty input", () => {
        const result = extractSimulationResults("");
        expect(result).toEqual({
            optimizedGeometry: null,
            minimizedEnergy: null,
            totalTime: null,
        });
    });

    test("handles partial results correctly", () => {
        const partialContent = `
            ================ OPTIMIZED GEOMETRY INFORMATION ==============
            OPTIMIZED GEOMETRY IN CARTESIAN
            ELEMENT X Y Z
            C -11.955388 4.422024 -0.357850
            C -9.634206 2.630167 0.650537
            
            FORCE
            MINIMIZED ENERGY = -840.694818961
        `.trim();

        const result = extractSimulationResults(partialContent);
        expect(result).toEqual({
            optimizedGeometry: "C -11.955388 4.422024 -0.357850\nC -9.634206 2.630167 0.650537",
            minimizedEnergy: -840.694818961,
            totalTime: null,
        });
    });

    test("correctly trims geometry output", () => {
        const inputWithSpaces = `
            OPTIMIZED GEOMETRY IN CARTESIAN
            ELEMENT X Y Z
              C    -11.955388    4.422024    -0.357850   
               C     -9.634206     2.630167      0.650537    
            FORCE
        `.trim();

        const result = extractSimulationResults(inputWithSpaces);
        expect(result.optimizedGeometry)
            .toBe("C    -11.955388    4.422024    -0.357850\nC     -9.634206     2.630167      0.650537");
    });

    test("correctly extracts minimized energy from test file", () => {
        const result = extractSimulationResults(successTestFileContent);
        expect(result.minimizedEnergy).toBe(-840.694818961);
    });

    test("correctly extracts total time from test file", () => {
        const result = extractSimulationResults(successTestFileContent);
        expect(result.totalTime).toBe(3967.181602000);
    });
});
