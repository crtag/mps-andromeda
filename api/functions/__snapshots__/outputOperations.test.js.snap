// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecule Parser Tests extractMoleculeInput extracts correct atom number and electrons from real test file 1`] = `
{
  "totalAtomNumber": 31,
  "totalElectrons": 62,
}
`;

exports[`extractSimulationResults extracts simulation results from real test file 1`] = `
{
  "minimizedEnergy": -840.694818961,
  "optimizedGeometry": "C         -11.955388       4.422024      -0.357850
C          -9.634206       2.630167       0.650537
C         -10.432765       2.739222       1.903391
C          -8.632891       3.685179       0.324675
C         -10.183896       1.744947      -0.412873
C         -11.679289       2.007138      -0.636453
H         -10.080590       0.686362      -0.130086
H          -9.654582       1.886622      -1.361639
C         -12.440840       1.888394       0.662417
C         -12.256556       1.113879      -1.727453
O         -11.836709       3.339530      -1.153021
O         -12.084589       4.237773       0.966049
C         -11.917430       2.953729       1.597211
H         -12.301251       0.889035       1.077761
H         -13.511066       2.040776       0.477665
H         -10.064242       3.566338       2.515032
H         -10.371613       1.830352       2.517116
C         -12.738068       3.051390       2.878696
H         -11.738892       1.324509      -2.671323
H         -13.311567       1.364161      -1.866447
O         -12.189809      -0.247624      -1.372352
H          -8.004295       3.354924      -0.516930
O          -7.848349       4.063972       1.436633
H          -9.123531       4.614098      -0.002815
H         -13.764521       3.341786       2.620081
H         -12.319127       3.834965       3.513950
O         -12.694299       1.852538       3.615400
H         -13.324566       1.235709       3.237484
H          -7.430399       3.268635       1.777838
H         -11.345260      -0.600844      -1.656667
O         -11.976763       5.514844      -0.828488",
  "totalTime": 3967.181602,
}
`;
